Connect Refresh: Use Calypso lost-password form for all unified Logins

chriskmnds opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of https://linear.app/a8c/issue/DOTCOM-13217/calypso-update-and-unify-the-lost-password-screen

## Proposed Changes

- This PR attempts to bring all Login flows/screens to utilise the Calypso lost-password form instead of the WP Admin one.
- Prior to this, only Woo and Blaze Pro clients would send users to this screen. Everything else redirects to the PHP admin page `/wp-login.php?action=lostpassword&redirect_to=...`

This PR is an attempt, but not something to merge if there are concerns. The Calypso form may not be seem as robust as the WP Admin one e.g. there is presumably missing information and links on errors, back links, it appears to work only with email addresses (per the errors shown). Compare the two below:

| Calypso (& on Error) | WP Admin (& on Error) |
|--------|--------|
| <img width="1041" alt="Screenshot 2025-06-20 at 4 53 18 PM" src="https://github.com/user-attachments/assets/175cea0c-b568-47a4-92be-ead1108fd0a2" /> | <img width="1041" alt="Screenshot 2025-06-20 at 4 53 57 PM" src="https://github.com/user-attachments/assets/62ab1e44-579a-426e-826a-a993d85401f7" /> | 
| <img width="1043" alt="Screenshot 2025-06-20 at 4 53 30 PM" src="https://github.com/user-attachments/assets/3d34ccc9-99e3-42d8-9dd8-cc16e71fbec8" /> | <img width="1037" alt="Screenshot 2025-06-20 at 4 54 13 PM" src="https://github.com/user-attachments/assets/81ee8c90-5602-4499-937d-e490e577fa7a" /> | 

I'm not sure if these are actual concerns, as the form is currently in use for Woo & Blaze. So not sure why there'd be issues elsewhere. Until recently, everything actually utilised the Calyspo form (so not just Woo or Blaze Pro). These related PR/issues added the redirection in the code (potentially others, but for reference): 
- https://github.com/Automattic/wp-calypso/issues/96291
- https://github.com/Automattic/wp-calypso/pull/98390

cc @escapemanuele @StevenDufresne, who worked on the above, can you please provide some insight?

With these changes, all client Logins will use the Calypso form:

| Akismet | A4A | JP Cloud (needs some style fixes) |
|--------|--------|--------|
| <img width="972" alt="Screenshot 2025-06-20 at 5 06 11 PM" src="https://github.com/user-attachments/assets/4f8b2ad8-3655-40da-9962-bcfe2b0b6fed" /> | <img width="961" alt="Screenshot 2025-06-20 at 5 07 00 PM" src="https://github.com/user-attachments/assets/225e68c9-1d65-4748-80db-41cea7f5a2c5" /> | <img width="958" alt="Screenshot 2025-06-20 at 5 06 49 PM" src="https://github.com/user-attachments/assets/a14b4954-e043-4c3c-a998-c2dffc02a688" /> |

### Realistic Alternative

For the sake of unification, an alternative might be to send all usages to WP Admin.

## TODO

- [x] Update the sub-title to not be TOS, but something else. Confirm.
- [ ] Bring back the footer. Decide what to link there. Confirm.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

Part of https://linear.app/a8c/issue/DOTCOM-13217/calypso-update-and-unify-the-lost-password-screen

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to any of the [client Logins](https://linear.app/a8c/issue/DOTCOM-13220/login-make-the-two-columnwhite-layout-the-default-across-all-contexts) (except Gravatar) and click on "lost your password" link at the bottom
* Confirm form renders and is functional
* **Confirm that it works as expected.** The lostpassword API works with WordPress.com, so you need to use a proxy here:
  * Proxy WordPress.com PCYsg-5YE-p2 to test the full flow and redirections
  * Alternative proxying method in https://github.com/Automattic/wp-calypso/pull/68558 (if the above fail)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

chriskmnds committed:
Connect Refresh: Unify lost-password screen

chriskmnds committed:
Connect Refresh: Use Calypso lost-password form for all unified Logins

chriskmnds commented on client/blocks/login/index.jsx:
```diff
@@ -444,29 +444,6 @@ class Login extends Component {
 							from={ get( currentQuery, 'from' ) }
 						/>
 					</div>
-					{ ! isWooJPC && ! isBlazePro && (
-						<div className="login__lost-password-footer">
-							<p className="login__lost-password-no-account">
-								{ translate( 'Don’t have an account? {{signupLink}}Sign up{{/signupLink}}', {
-									components: {
-										signupLink,
-									},
-								} ) }
-							</p>
-						</div>
-					) }
-					{ isBlazePro && (
-						<div className="login__lost-password-footer">
-							<p className="login__lost-password-no-account">
-								<span>{ translate( 'Don’t have an account?' ) }&nbsp;</span>
-								{ translate( '{{signupLink}}Sign up{{/signupLink}}', {
-									components: {
-										signupLink,
-									},
-								} ) }
-							</p>
-						</div>
-					) }
```
My mistake. This should be brought back if proceeding, but unified.

chriskmnds replied:
No mistake. Footer goes through a different component.

chihsuan reviewed (commented):
The Calypso lost-password screen was added during the Woo.com onboarding (NUX) improvement project.

It was introduced for Woo only at that time because the WP Admin lost password page lacks design consistency with Woo pages. Also, it's WP-branded, and we need a Woo-branded one.

Later, Blaze Pro also started using the Calypso form when they created their auth flow.

There were some unexpected issues when someone changes the code without awareness that different products use different lost password screens. So I'm glad to see this PR. Thanks for working on this! 🙏 🙂 

Just a FYI, this is the design for Woo lost password screen. If we wanted to change the code, it would be good to reference this design.

<img width="700" alt="Screenshot 2025-06-23 at 14 27 11" src="https://github.com/user-attachments/assets/5ec1714b-e382-41bb-90a4-d208598655f2" />


tp2u3Nk9arcpeNMZcylNG4-fi-616_4610

chriskmnds committed:
update subheading text to match design

chriskmnds commented:
Thanks @chihsuan ! 

> There were some unexpected issues when someone changes the code without awareness that different products use different lost password screens. 

Wonderful. Since we are unifying all the screens now (except Gravatar - which will continue with the old form), then this won't be an issue. If this has been working fine for Woo and Blaze, then we should be able to adapt throughout. :-)

I'd love to hear from @escapemanuele and @StevenDufresne as well if possible, in case they had other concerns/issues with this form.

> Just a FYI, this is the design for Woo lost password screen. If we wanted to change the code, it would be good to reference this design.

Thanks. I will be updating the subtext to match the QwLWndRBggFq0keb79EZM7-fi-2280_589, which will read: `Please enter your username or email address. You'll receive a link to create a new password via email.`.

escapemanuele commented:
I don't have particular concerns here, I really hope we can proceed with this soon.

I tried to use my personal email, and it gave me error because it couldn't find http://calypso.localhost:3000/wp-login.php?action=lostpassword, probably I should have mocked wordpress.com

![image](https://github.com/user-attachments/assets/6193923b-4e6a-4db7-8b71-683d19af02a2)

But the thing is that I'm not able to try again, even if I cancel the email and try to put another one. The error persists and the button remains disabled.

chriskmnds committed:
useSignupUrl

chriskmnds committed:
cleanup. shouldRenderSignupUrl -> false

chriskmnds commented:
Cleaned up. Design-wise, it is now without any bottom links:

<img width="500" alt="Screenshot 2025-06-23 at 7 03 51 PM" src="https://github.com/user-attachments/assets/b5364d55-3479-4796-9df1-08834f5ba446" />

The signup nudge is already covered by "create an account" from the header that's always visible. I moved the code around to render it, but forced it to `false`. will clean up once this is better tested and ready.

> But the thing is that I'm not able to try again, even if I cancel the email and try to put another one. The error persists and the button remains disabled.

Thanks for confirming @escapemanuele . This is the behavior in production too, and it is indeed weird. We should be able to try more than once. I will try and work it out, but cc @chihsuan who worked on this initially. Was the reason we disable the button completely after the first attempt/error?

chriskmnds committed:
cleanup. remove useSignupLinkComponent

chriskmnds committed:
remove isLostPassword

chriskmnds commented on client/login/wp-login/index.jsx:
```diff
@@ -606,7 +567,7 @@ export class Login extends Component {
 									<HeadingSubText
 										isSocialFirst={ isSocialFirst }
 										twoFactorAuthType={ twoFactorAuthType }
-										isLostPassword={ isLostPassword }
```
@raicem Following up on our chat https://github.com/Automattic/wp-calypso/pull/104104#discussion_r2161059772 - do you mind testing again with the Woo flows if this works? I think it works for me, but I may not be triggering the WooJPC flows correctly or something. 🙏

raicem replied:
Can confirm it works now for me, I wonder what I got wrong on Monday. Thank you for this @chriskmnds! 

I checked WooCommerce.com and WooJPC login and password reset. Their styles are consistent, so looks good! I'm happy to review it more thoroughly for other flows too, if needed.

jboland88 commented on client/login/wp-login/login-footer.tsx:
```diff
@@ -1,42 +1,72 @@
 	const oauth2Client = useSelector( getCurrentOAuth2Client );
 	const isVIPClient = isVIPOAuth2Client( oauth2Client );
+	const signupUrl = useSignupUrl( { signupUrl: signupUrlProp } );
+	const shouldRenderSignupLink = 'lostpassword' === action && false;
```
I was wondering why this was set to always be `false`, I see you mentioned it in this comment: https://github.com/Automattic/wp-calypso/pull/104367#issuecomment-2997074217

chriskmnds replied:
Exactly, yeah

jboland88 commented on client/login/wp-login/index.jsx:
```diff
@@ -273,58 +273,23 @@ export class Login extends Component {
-			const currentUrl = new URL( window.location.href );
-			currentUrl.searchParams.append( 'lostpassword_flow', true );
-			const queryArgs = {
-				redirect_to: currentUrl.toString(),
```
Is this redirect handled in another way after this update?

chriskmnds replied:
It should be handled like all the others through `redirectTo` passed to `login`.
The only difference is why this was mapped to `window.location.href` and not `redirectTo` grabbed from state below / whether that's intentional 

@escapemanuele do you recall off-hand?

jboland88 replied:
I tested the JP cloud login and got redirected back to cloud.jetpack.com, so, seems to work as expected 🤷‍♂️

jboland88 reviewed (changes requested):
I tested with the WPCOM and Jetpack Cloud logins to see how this was working.

For WPCOM in particular, I would consider the lack of support for username on the password recovery screen a blocking regression. Is that something we could add to the screen separately first before this merges?

When testing Jetpack cloud, I noticed an odd styling issue where the field width was small before submitting an email, but wide in the error/ complete state

![Screenshot 2025-06-23 at 4 59 02 PM](https://github.com/user-attachments/assets/33252942-df1a-4aa1-9644-bbc994182f4a)
![Screenshot 2025-06-23 at 4 59 15 PM](https://github.com/user-attachments/assets/b1011258-b01a-4955-90e2-6ab72cb54282)

I was also not able to successfully submit a request to reset from the JP Cloud login, It always provided an error. I was using an @automattic.com+modifier address for a test account. I tried the same email address with the current wp-admin password recovery form and it worked correctly 🤔

chriskmnds committed:
remove unused

chriskmnds committed:
remove unused

chriskmnds committed:
cleanup login form style overrides

jboland88 reviewed (approved):
I tested with a local proxy and this is working as expected for me.
I did not send a reset request with every service (since the use the same routine), but I visually checked all the services with updated login screens.

I noticed a few small improvements that could be handled here or separately, no blockers.

Not a blocker, but I think we should re-add the "Log-in" link below the input to this screen similar to the current wp-login.php?action=lostpassword screen so that the user has a clear way to get back to the login screen without hitting the browser back button. Can be handled separately.

[Jetpack ](https://wordpress.com/log-in/jetpack) shows the generic WPCOM forgot password screen. It would be great if the same Jetpack branding used for the Jetpack Cloud forgot password screen could be used there as well. I'm guessing the branding may somehow be connected with the oauth client?
![Screenshot 2025-06-25 at 5 23 34 PM](https://github.com/user-attachments/assets/43a56484-3387-4a2d-9f1f-aab02a7141b3)

Nitpick: the input field width on the Woo and Blaze pro screens are slightly narrower than the others.
![Screenshot 2025-06-25 at 5 23 55 PM](https://github.com/user-attachments/assets/5fdc1f2b-3529-4ae8-9c5a-ce14d18451be)
![Screenshot 2025-06-25 at 5 26 27 PM](https://github.com/user-attachments/assets/a7aa7a19-e8ca-4705-b2d5-6b4c17e8e7e5)

This is a design nitpick, but I noticed that we don't have the name of the service you are logging into anymore on the confirmation screen. In some cases the logo is different (Like Akismet), but in the case of Studio or Jetpack- there is no branding and sometimes no brand colors anymore (for non-oauth logins). (Second screenshot is of the Jetpack log-in). This seems to work fine for the Woo onboarding logins though, so perhaps a query param is sufficient.
![Screenshot 2025-06-25 at 5 28 22 PM](https://github.com/user-attachments/assets/61c2b3c6-3b1f-43d3-88ca-1063ef9d8d0d)
![Screenshot 2025-06-25 at 5 32 40 PM](https://github.com/user-attachments/assets/926c68fe-0aeb-4f9b-9692-473549866547)
![Screenshot 2025-06-25 at 5 35 38 PM](https://github.com/user-attachments/assets/bd2779bd-cd7b-404d-800c-8171fd6924cc)

jboland88 commented:
Oh, one other thing that you've already noted somewhere else @chriskmnds - It is quite annoying that the submit button does not get re-enabled after a failed submission unless you move focus out of the input field.

I think having any input change - like re-typing the username or email - should also clear that disabled state and remove the error. cc @chihsuan since I think this may be something already on the radar?

chriskmnds commented:
Thanks again @jboland88. Solid points, all worth addressing.

> Not a blocker, but I think we should re-add the "Log-in" link below the input to this screen similar to the current `wp-login.php?action=lostpassword` screen so that the user has a clear way to get back to the login screen without hitting the browser back button. Can be handled separately.

I do agree. We may be adding a "back" button in the top bar, which would behave similarly, but no harm in having a direct "Log in" CTA in the bottom.

Concerning `wp-login.php?action=lostpassword`, it's actually always sending to the default `/log-in`, so it might even be breaking the flow/branding for client apps. I think more pronounced in the case of Gravatar. So we'll try and add a proper customised redirect instead.
 
> [Jetpack ](https://wordpress.com/log-in/jetpack) shows the generic WPCOM forgot password screen. It would be great if the same Jetpack branding used for the Jetpack Cloud forgot password screen could be used there as well. I'm guessing the branding may somehow be connected with the oauth client?

Indeed. I confirm the same:

<img width="500" alt="Screenshot 2025-06-26 at 6 45 19 PM" src="https://github.com/user-attachments/assets/538c3313-f840-487b-8217-132c75abb872" />

However, I think the issue is with the links and redirections. For JP it should be sending to `/log-in/jetpack/lostpassword` and from there back to `/log-in/jetpack?email_address=xxx&lostpassword_flow=true`:

| `/log-in/jetpack/lostpassword` | `/log-in/jetpack?lostpassword_flow=true` |
|--------|--------|
| <img width="500" alt="Screenshot 2025-06-26 at 6 57 51 PM" src="https://github.com/user-attachments/assets/dcb95979-b59c-4b5c-b6b4-89bc8043c5b3" /> | <img width="840" alt="Screenshot 2025-06-26 at 7 01 16 PM" src="https://github.com/user-attachments/assets/316c3220-b97d-448e-a7ef-a5bc54643ba2" /> | 

I'll try and work these in; otherwise, separate issues and do on top.

chriskmnds committed:
fix JP redirects

chriskmnds commented on client/login/wp-login/index.jsx:
```diff
@@ -272,58 +272,26 @@ export class Login extends Component {
+						login( {
+							redirectTo: this.props.redirectTo,
+							locale: this.props.locale,
+							action:
+								this.props.isWooJPC || this.props.isJetpack
+									? 'jetpack/lostpassword'
+									: 'lostpassword',
```
@raicem @chihsuan Can you help me connect again why `WooJPC` sends to `/log-in/jetpack/lostpassword` and not `/log-in/lostpassword`? Or is this something we can clean up now with the unification (everything using the Calypso page)?

chihsuan replied:
This was introduced in [this commit](https://github.com/Automattic/wp-calypso/commit/ead9fba208d5cd8c13489b27f6b3dd81047b6b2b) for WooJPC forgot password functionality, but appears to no longer be in active use.

What's the plans for Woo JPC screens? If we're planning to standardize them, then removal would be appropriate.

chriskmnds replied:
They are being (have been) standardised like the rest. @raicem worked on that. 
I'll need to check these again and confirm we don't lose the branding. There are a few methods to trigger these WooJPC flows from what I recall.

chriskmnds committed:
back to login links

chriskmnds commented:
Fixed the JP redirections and added the "Back to Login" link. 
@jboland88 If you'd like to take another look, in case I missed anything.
Also added a focus now on the lost-password form input cc @chihsuan - again, in case anything I missed anythning there (whether not being focused was intentional) 🤔 


https://github.com/user-attachments/assets/ca66de54-18d5-4c0b-aa92-d2b557afde3e

> Also added a focus now on the lost-password

We'll have to fix the focus style for the lost-password fields (you can see it's different from the main Login, for example, in JP's case)

chriskmnds commented:
> > Also added a focus now on the lost-password
> 
> We'll have to fix the focus style for the lost-password fields (you can see it's different from the main Login, for example, in JP's case)

Looks like we have an issue for this already: https://linear.app/a8c/issue/DSGCOM-109/update-text-fields-on-lost-password

chriskmnds commented:
Let's go! Thanks for the reviews, everyone 🎉
